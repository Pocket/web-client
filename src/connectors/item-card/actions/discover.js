import React from 'react'
import { ItemActions } from 'components/item-menus/flat'
import { SaveIcon } from '@pocket/web-ui'
import { ReportIcon } from '@pocket/web-ui'

export function ActionsDiscover({ id, isAuthenticated, save_status }) {
  const onSaveClick = () => {}
  const onReportFeedback = () => {}

  return (
    <div className="actions">
      <ItemActions
        menuItems={[
          {
            label: 'Save To Pocket',
            icon: <SaveIcon />,
            onClick: onSaveClick
          }
        ]}
        overflowItems={[
          {
            label: 'Report',
            actionText: 'Report',
            icon: <ReportIcon />,
            onClick: onReportFeedback
          }
        ]}
      />
    </div>
  )
}
