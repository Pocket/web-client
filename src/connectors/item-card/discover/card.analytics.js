import { trackImpression } from 'connectors/snowplow/snowplow.state'
import { trackContentOpen } from 'connectors/snowplow/snowplow.state'
import { CONTENT_OPEN_TRIGGER_CLICK } from 'connectors/snowplow/events'
import { IMPRESSION_COMPONENT_CARD } from 'connectors/snowplow/events'
import { IMPRESSION_REQUIREMENT_VIEWABLE } from 'connectors/snowplow/events'
import { getLinkOpenTarget } from 'connectors/snowplow/events'

/**
 * fireItemImpression - function to conditionally execute when an item is
 * in view on the page. Only fires if an impressionAction is passed in.
 * @param {int} position position of card on page
 * @param {int} positionZeroIndex position of card on page, 0-indexed
 * @param {object} item rendered by card
 * @param {bool} inView is item in view
 * @param {function} impressionAction analytics action for impression
 */
export function fireItemImpression(
  position,
  positionZeroIndex,
  item,
  inView,
  impressionAction,
  dispatch
) {
  if (inView && impressionAction) {
    impressionAction(position, item)

    // trigger Snowplow impression
    dispatch(
      trackImpression(
        IMPRESSION_COMPONENT_CARD,
        IMPRESSION_REQUIREMENT_VIEWABLE,
        positionZeroIndex,
        item,
        'web-discover-card'
      )
    )
  }
}

export function fireItemOpen(positionZeroIndex, item, dispatch) {
  const linkTarget = getLinkOpenTarget(item?.save_url, item?.syndicated)
  // trigger Snowplow content open
  dispatch(
    trackContentOpen(
      linkTarget,
      CONTENT_OPEN_TRIGGER_CLICK,
      positionZeroIndex,
      item
    )
  )
}
