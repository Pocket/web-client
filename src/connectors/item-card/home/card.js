import { useEffect } from 'react'
import { Card } from 'components/item-card/home/topic-card'
import { useSelector, useDispatch } from 'react-redux'
import { useInView } from 'react-intersection-observer'

export const ItemCard = ({ id, topic, unSaveAction, saveAction, position }) => {
  // Get data from state
  const isAuthenticated = useSelector((state) => state.user.auth)
  const item = useSelector((state) => state.home.topics[topic]?.itemsById[id])

  const { save_url, save_status } = item
  const dispatch = useDispatch()

  // Fire item impression
  const [ref, inView] = useInView({ triggerOnce: true, threshold: 0.5 })
  useEffect(() => {
    // track impression event
    if (inView) console.log('impression')
  }, [inView])

  const onSave = (isAuthenticated) => {
    if (isAuthenticated) {
      if (save_status === 'saved') dispatch(unSaveAction(id, topic))
      if (save_status !== 'saved') dispatch(saveAction(id, topic, save_url, position))
      return
    }

    // Not authenticated so just tracking the click
    // unAuthSaveAction(position)
  }

  const onOpen = () => {
    // track open event
  }

  return (
    <Card
      ref={ref}
      item={item}
      onOpen={onOpen}
      onSave={onSave}
      isAuthenticated={isAuthenticated}
    />
  )
}
