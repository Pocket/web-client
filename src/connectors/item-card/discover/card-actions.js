import React from 'react'
import { ReportIcon } from 'components/icons/ReportIcon'
import { SaveToPocket } from 'components/item-actions/save-to-pocket'
import { OverflowAction } from 'components/item-actions/overflow'
import { useSelector, useDispatch } from 'react-redux'
import { saveDiscoverItem } from 'containers/discover/discover.state'
import { useTranslation } from 'next-i18next'

import { itemActionStyle } from 'components/item-actions/base'
import { itemReportAction } from 'connectors/items/mutation-report.state'

import { sendSnowplowEvent } from 'connectors/snowplow/snowplow.state'

export function ActionsDiscover({ id, position }) {
  const dispatch = useDispatch()
  const { t } = useTranslation()

  const isAuthenticated = useSelector((state) => state.user.auth)
  const item = useSelector((state) => state.discoverItemsById[id])

  if (!item) return null
  const { saveStatus, saveUrl, externalUrl, openExternal } = item
  const analyticsData = { position, ...item?.analyticsData }

  // Prep save action
  const onSave = () => {
    dispatch(saveDiscoverItem(id, saveUrl, position))
    dispatch(sendSnowplowEvent('discover.save', analyticsData))
  }

  // Open action
  const url = externalUrl

  const onOpen = () => {
    const data = {
      ...analyticsData,
      url,
      destination: openExternal ? 'external' : 'internal'
    }
    dispatch(sendSnowplowEvent('discover.open', data))
  }

  // On Report
  const onReport = () => dispatch(itemReportAction(id))

  return item ? (
    <div className={`${itemActionStyle} actions`}>
      <SaveToPocket
        allowRead={false}
        url={url}
        onOpen={onOpen}
        openExternal={openExternal}
        saveAction={onSave}
        isAuthenticated={isAuthenticated}
        saveStatus={saveStatus}
        id={id}
      />

      <OverflowAction
        menuItems={[
          {
            label: t('item-action:report', 'Report'),
            actionText: t('item-action:report', 'Report'),
            icon: <ReportIcon />,
            onClick: onReport
          }
        ]}
      />
    </div>
  ) : null
}
