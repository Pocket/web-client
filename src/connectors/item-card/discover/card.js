import { Card } from 'components/item-card/card'
import { useSelector, useDispatch } from 'react-redux'
import { ActionsDiscover } from './card-actions'
import { setNoImage } from 'connectors/items-by-id/discover/items.state'
import { trackItemImpression } from 'connectors/snowplow/snowplow.state'
import { trackItemOpen } from 'connectors/snowplow/snowplow.state'

/**
 * Article Card
 * Creates a connected `Card` with the appropriate data and actions
 * @param {object} {id, position} item_id for data and position for analytics
 */
export function ItemCard({ id, position, className, cardShape, showExcerpt = false }) {
  const dispatch = useDispatch()

  // Get data from state
  const item = useSelector((state) => state.discoverItemsById[id])
  const { save_status, item_id, original_url, openExternal } = item
  const impressionFired = useSelector((state) => state.analytics.impressions.includes(id))
  const openUrl = save_status === 'saved' && !openExternal ? `/read/${item_id}` : original_url
  const onImageFail = () => dispatch(setNoImage(id))

  /**
   * ITEM TRACKING
   * ----------------------------------------------------------------
   */
  const onImpression = () => dispatch(trackItemImpression(position, item, 'discover.impression'))
  const onItemInView = (inView) => (!impressionFired && inView ? onImpression() : null)
  const onOpen = () => dispatch(trackItemOpen(position, item, 'discover.open'))

  return (
    <Card
      item={item}
      onImageFail={onImageFail}
      position={position}
      className={className}
      cardShape={cardShape}
      showExcerpt={showExcerpt}
      openUrl={openUrl}
      // Tracking
      onItemInView={onItemInView}
      openAction={onOpen}
      ActionMenu={ActionsDiscover}
    />
  )
}
