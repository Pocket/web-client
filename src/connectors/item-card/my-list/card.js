import { memo } from 'react'
import { Card } from 'components/item-card/my-list/card'
import { useSelector, useDispatch } from 'react-redux'

import { itemsArchiveAction } from 'connectors/items-by-id/my-list/items.archive'
import { itemsUnArchiveAction } from 'connectors/items-by-id/my-list/items.archive'

import { itemsFavoriteAction } from 'connectors/items-by-id/my-list/items.favorite'
import { itemsUnFavoriteAction } from 'connectors/items-by-id/my-list/items.favorite'

import { itemsDeleteAction } from 'connectors/items-by-id/my-list/items.delete'
import { itemsTagAction } from 'connectors/items-by-id/my-list/items.tag'
import { itemsShareAction } from 'connectors/items-by-id/my-list/items.share'

import { itemsBulkSelectAction } from 'connectors/items-by-id/my-list/items.bulk'
import { itemsBulkDeSelectAction } from 'connectors/items-by-id/my-list/items.bulk'

import { fireItemOpen } from 'connectors/items-by-id/my-list/items.analytics'
import { trackItemOpen } from 'connectors/items-by-id/my-list/items.analytics'
import { setImpression } from 'connectors/items-by-id/my-list/items.analytics'
import { sendEngagementEvent } from 'connectors/items-by-id/my-list/items.analytics'
import { selectShortcutItem } from 'connectors/shortcuts/shortcuts.state'

import copy from 'clipboard-copy'
import { COPY_ITEM_URL } from 'actions'

/**
 * Article Card
 * Creates a connected `Card` with the appropriate data and actions
 * @param {object} {id, position} item_id for data and position for analytics
 */
export function ItemCard({ id, position, fluidHeight, type }) {
  const dispatch = useDispatch()

  const appMode = useSelector((state) => state?.app?.mode)
  const bulkEdit = appMode === 'bulk'

  const isPremium = useSelector((state) => state.user.premium_status === '1')

  // Get data from state
  const item = useSelector((state) => state.myListItemsById[id])
  const impression = useSelector((state) => state.itemsAnalytics.impressions)

  const bulkList = useSelector((state) => state.bulkEdit.selected)
  const bulkCurrent = useSelector((state) => state.bulkEdit.currentId)

  const bulkIsCurrent = bulkCurrent === id
  const bulkSelected = bulkList?.map((item) => item.id).includes(id)

  const shortcutId = useSelector((state) => state.shortcuts.currentId)
  const shortcutSelected = shortcutId === id

  const itemImpression = () => {
    if (!impression[item.item_id]) {
      dispatch(setImpression(position, item))
    }
  }

  const itemShare = () => {
    dispatch(sendEngagementEvent('my-list.share', position, item))
    dispatch(itemsShareAction({ id, position }))
  }
  const itemDelete = () => {
    dispatch(sendEngagementEvent('my-list.delete', position, item))
    dispatch(itemsDeleteAction([{ id, position }]))
  }

  const itemArchive = () => {
    dispatch(sendEngagementEvent('my-list.archive', position, item))
    dispatch(itemsArchiveAction([{ id, position }]))
  }
  const itemUnArchive = () => {
    // bool to denote save action
    dispatch(sendEngagementEvent('my-list.unarchive', position, item, true))
    dispatch(itemsUnArchiveAction([{ id, position }]))
  }

  const itemFavorite = () => {
    dispatch(sendEngagementEvent('my-list.favorite', position, item))
    dispatch(itemsFavoriteAction([{ id, position }]))
  }
  const itemUnFavorite = () => {
    dispatch(sendEngagementEvent('my-list.un-favorite', position, item))
    dispatch(itemsUnFavoriteAction([{ id, position }])) //prettier-ignore
  }

  const itemTag = () => {
    dispatch(sendEngagementEvent('my-list.tag', position, item))
    dispatch(itemsTagAction([{ id, position }]))
  }

  const itemBulkSelect = (shift) => dispatch(itemsBulkSelectAction(id, shift)) //prettier-ignore
  const itemBulkDeSelect = (shift) => dispatch(itemsBulkDeSelectAction(id, shift)) //prettier-ignore

  const copyAction = () => ({ type: COPY_ITEM_URL })
  const itemCopy = async () => {
    await copy(item?.open_url)
    dispatch(copyAction())
  }

  const onOpen = () => {
    trackItemOpen(position + 1, item, type) // legacy analytics uses 1 based position
    fireItemOpen(position, item, dispatch)
  }

  const shortcutSelect = () => dispatch(selectShortcutItem(id, position))

  return item ? (
    <Card
      item={item}
      position={position}
      fluidHeight={fluidHeight}
      type={type}
      bulkEdit={bulkEdit}
      bulkSelected={bulkSelected}
      bulkIsCurrent={bulkIsCurrent}
      shortcutSelected={shortcutSelected}
      shortcutSelect={shortcutSelect}
      onOpen={onOpen}
      actions={{
        itemShare,
        itemDelete,
        itemArchive,
        itemUnArchive,
        itemFavorite,
        itemUnFavorite,
        itemTag,
        itemBulkSelect,
        itemBulkDeSelect,
        itemImpression,
        itemCopy
      }}
      isPremium={isPremium}
    />
  ) : null
}

// This seems like an over-optimization so do some actual testing here
const isEqual = (prev, next) => {
  const isTheSame =
    prev.id === next.id &&
    prev.position === next.position &&
    prev.type === next.type
  return isTheSame
}

export const MemoizedItem = memo(ItemCard, isEqual)
