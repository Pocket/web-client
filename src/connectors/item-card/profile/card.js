import { Card } from 'components/item-card/card'
import { useSelector, useDispatch } from 'react-redux'
import { ActionsFeed } from './card-actions'
import { PostHeader } from 'components/post-header/post-header'
import { sendSnowplowEvent } from 'connectors/snowplow/snowplow.state'

export const RecommendedFeedCard = ({ id, position }) => {
  const dispatch = useDispatch()

  // Get data from state
  const impressionFired = useSelector((state) => state.analytics.impressions.includes(id))
  const item = useSelector((state) => state.profileItemsByIds.itemsById[id])

  if (!item) return null

  const { readUrl, externalUrl, analyticsData: passedAnalytics, openExternal } = item
  const openUrl = readUrl || externalUrl
  const itemImage = item?.noImage ? '' : item?.thumbnail
  const {title, publisher, excerpt, timeToRead, isSyndicated, post} = item //prettier-ignore
  const analyticsData = {
    id,
    position,
    ...passedAnalytics
  }

  /**
   * ITEM TRACKING
   * ----------------------------------------------------------------
   */
  const onOpen = () => dispatch(sendSnowplowEvent('profile.feed.open', analyticsData))
  const onImpression = () => dispatch(sendSnowplowEvent('profile.feed.impression', analyticsData))
  const onItemInView = (inView) => (!impressionFired && inView ? onImpression() : null)

  return item ? (
    <div className="recommendedWrapper">
      <PostHeader {...post} />
      <Card
        itemId={id}
        title={title}
        publisher={publisher}
        excerpt={excerpt}
        timeToRead={timeToRead}
        isSyndicated={isSyndicated}
        itemImage={itemImage}
        externalUrl={externalUrl}
        openExternal={openExternal}
        cardShape="wide"
        position={position}
        showExcerpt={true}
        onItemInView={onItemInView}
        onOpen={onOpen}
        onOpenOriginalUrl={onOpen}
        hiddenActions={false}
        openUrl={openUrl}
        ActionMenu={ActionsFeed}
      />
    </div>
  ) : null
}
