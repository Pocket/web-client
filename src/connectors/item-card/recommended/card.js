import { css } from 'linaria'
import { Card } from 'components/item-card/card'
import { useSelector, useDispatch } from 'react-redux'
import { ActionsFeed } from './card-actions'
import { PostHeader } from 'components/post-header/post-header'

import { trackItemImpression } from 'connectors/snowplow/snowplow.state'
import { trackItemOpen } from 'connectors/snowplow/snowplow.state'

export const RecommendedFeedCard = ({ id, position }) => {
  const dispatch = useDispatch()

  // Get data from state
  const impressionFired = useSelector((state) => state.analytics.impressions.includes(id))
  const item = useSelector((state) => state.recommendedItemsByIds.itemsById[id])

  const { save_status, item_id, original_url, openExternal, post } = item
  const openUrl = save_status === 'saved' && !openExternal ? `/read/${item_id}` : original_url

  /**
   * ITEM TRACKING
   * ----------------------------------------------------------------
   */
  const onOpen = () => dispatch(trackItemOpen(position, item, 'recommended.feed.open'))
  const onImpression = () => dispatch(trackItemImpression(position, item, 'recommended.feed.impression'))
  const onItemInView = (inView) => (!impressionFired && inView ? onImpression() : null)

  return item ? (
    <div className="recommendedWrapper">
      <PostHeader {...post} />
      <Card
        id={id}
        cardShape="wide"
        position={position}
        item={item}
        showExcerpt={true}
        onItemInView={onItemInView}
        onOpen={onOpen}
        onOpenOriginalUrl={onOpen}
        hiddenActions={false}
        openUrl={openUrl}
        ActionMenu={ActionsFeed}
      />
    </div>
  ) : null
}
