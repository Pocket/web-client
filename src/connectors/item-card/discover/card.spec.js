import { wrappedRender } from 'test-utils'
import '@testing-library/jest-dom/extend-expect'
import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils'

import { lineupAnalytics } from 'common/api/derivers/discover/discover.spec'
import { slateAnalytics } from 'common/api/derivers/discover/discover.spec'
import { recommendationsFromSlate } from 'common/api/derivers/discover/discover.spec'

import { ItemCard as DiscoverCard } from 'connectors/item-card/discover/card'
import { deriveDiscoverItems } from 'connectors/items-by-id/discover/items.derive'

describe('ItemCard', () => {
  //Legacy Derivers
  const derivedDiscoverItem = deriveDiscoverItems([recommendationsFromSlate])[0]

  const initialState = {
    user: { auth: true },
    analytics: {
      impressions: []
    },
    discoverItemsById: {
      [derivedDiscoverItem.item_id]: derivedDiscoverItem
    }
  }

  beforeAll(() => {
    mockAllIsIntersecting()
  })

  // Discover card
  it('renders a discover item', () => {
    const { getByCy } = wrappedRender(<DiscoverCard id={derivedDiscoverItem.item_id} position={3} />, { initialState }) //prettier-ignore
    expect(getByCy('article-card-', { exact: false })).toMatchSnapshot()
  })
})
