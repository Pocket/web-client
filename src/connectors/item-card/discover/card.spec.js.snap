// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemCard renders a discover item 1`] = `
<article
  class="c18o9ext grid noExcerpt"
  data-cy="article-card-3460462323"
>
  <div
    class="selectedBack"
  />
  <div
    class="cardWrap"
  >
    <div
      class="c97c5c media"
    >
      <a
        data-cy="image-link"
        href="http://theverge.com/22734645/apple-macbook-pro-2021-ports-magsafe-touch-bar-usb-c-future?utm_source=pocket_discover"
        tabindex="-1"
      >
        <img
          alt=""
          src="https://pocket-image-cache.com/600x400/filters:format(jpg):extract_focal()/https%3A%2F%2Fcdn.vox-cdn.com%2Fthumbor%2F26f5VRT9jvt6JW6WB4Nohu849cI%3D%2F0x137%3A1911x1138%2Ffit-in%2F1200x630%2Fcdn.vox-cdn.com%2Fuploads%2Fchorus_asset%2Ffile%2F22938855%2FApple_MacBook_Pro_Ports_10182021.jpg"
          srcset=" https://pocket-image-cache.com/1200x800/filters:format(jpg):extract_focal()/https%3A%2F%2Fcdn.vox-cdn.com%2Fthumbor%2F26f5VRT9jvt6JW6WB4Nohu849cI%3D%2F0x137%3A1911x1138%2Ffit-in%2F1200x630%2Fcdn.vox-cdn.com%2Fuploads%2Fchorus_asset%2Ffile%2F22938855%2FApple_MacBook_Pro_Ports_10182021.jpg 2x "
          style="--fallbackBackground: #00CB7780; --fallbackColor: #00CB77; --fallbackLetter: 'A';"
        />
      </a>
    </div>
    <div
      class="content"
    >
      <h2
        class="title"
      >
        <a
          data-cy="title-link"
          href="http://theverge.com/22734645/apple-macbook-pro-2021-ports-magsafe-touch-bar-usb-c-future?utm_source=pocket_discover"
          tabindex="0"
        >
          Apple is ready to admit it was wrong about the future of laptops
        </a>
      </h2>
      <cite
        class="details"
      >
        <a
          class="publisher"
          data-cy="publisher-link"
          href="http://theverge.com/22734645/apple-macbook-pro-2021-ports-magsafe-touch-bar-usb-c-future?utm_source=pocket_discover"
          tabindex="0"
        >
          The Verge
        </a>
        <span
          class="readtime"
          data-cy="read-time"
        >
           
          Â· 
          7
           min
        </span>
      </cite>
    </div>
  </div>
  <footer
    class="footer"
  >
    <div
      class="i18uycg6 actions"
    >
      <button
        class="b5bt6fr sb4ns5w unsaved card-actions"
        data-cy="article-save-btn-3460462323"
      >
        <span
          class="i1qqph0t icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M2 5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2H2Zm2 0H2v6c0 5.523 4.477 10 10 10s10-4.477 10-10V5h-2v6a8 8 0 1 1-16 0V5Z"
              fill-rule="evenodd"
            />
            <path
              clip-rule="evenodd"
              d="M7.293 9.293a1 1 0 0 1 1.414 0L12 12.586l3.293-3.293a1 1 0 1 1 1.414 1.414l-4 4a1 1 0 0 1-1.414 0l-4-4a1 1 0 0 1 0-1.414Z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <div
          class="actionCopy"
        >
          item-action:save-unsaved
        </div>
      </button>
      <button
        class="m11fpiro"
        data-cy="overflow"
      >
        <span
          class="i1qqph0t icon"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14 12a2 2 0 1 1-4 0 2 2 0 0 1 4 0ZM6 12a2 2 0 1 1-4 0 2 2 0 0 1 4 0ZM22 12a2 2 0 1 1-4 0 2 2 0 0 1 4 0Z"
            />
          </svg>
        </span>
      </button>
    </div>
  </footer>
</article>
`;
